### ISSUE TITLE
Route for subdomains

### ISSUE THREAD

**Issue by alikzao** [aio-libs/aiohttp#1342]
## Describe

Sometimes there are situations when it is necessary for the current project to create a subdomain just as another rout.
And hang on to this subdomain a handler that is already there.
## Expected behaviour

For example it might look something like this:

```
app.router.add_route('GET',  '/',  main_page)
app.router.add_route('GET',  '/',  handler, subdomain="domain")

def handler():
      print('text')
```

**Comment by asvetlov** [comment]
After #1301 it's possible to register subapplications by prefix.

I suggest the same technique but for resolving by subdomain.

**Comment by cynecx** [comment]
I am currently using the approach of adding a custom middleware which specifically filters out unknown hostnames, the only issue I am currently having is that it quite hard to support real micro-services with different domains, but this isn't an issue anymore especially after @asvetlov has landed sub-applications :).

**Comment by asvetlov** [comment]
I mean we could add a method `router.add_subdomain('sub.domain.com', subapp)` like recently added `router.add_subapp('/prefix', subapp)`.

**Comment by cynecx** [comment]
@asvetlov Do you also consider supporting wildcards in the domain name as well a falback so unknown domains are handled too?

**Comment by asvetlov** [comment]
Why not?

**Comment by cynecx** [comment]
@asvetlov Well, by looking at your example I thought you would like to keep it very simple, so in that case I assumed it would come without wildcard and fallback domain support. But yeah, it's good to know that you will include that too :).

**Comment by ghost** [comment]
Is it implemented?

**Comment by fafhrd91** [comment]
no, it is not

**Comment by ghost** [comment]
Very sad :(

**Comment by fafhrd91** [comment]
Would you provide PR?

**Comment by ghost** [comment]
Nah, I'm not familiar with aiohttp

**Comment by alikzao** [comment]
@TiberiumPY I did not wait for decisions, so switched to another framework

**Comment by ghost** [comment]
@alikzao what framework?

**Comment by fafhrd91** [comment]
You have plenty of options. Pyramid, flask.

**Comment by espositofulvio** [comment]
If this has not been done yet, I'd like to have add_domain instead of add_subdomain. It would be a nice feature which could easily enable multi-site support for aiohttp applications. Further, having remove_subapp() and remove_domain() would complete the feature allowing to dinamically add and remove features to a website and in my opinion could be a nice thing to have (paid features come to mind but also a multi-site cms where user can actually load the features they want could be handled this way easily).

**Comment by alex-eri** [comment]
I suggest to rewrite domain to prefix on front web-server (like nginx)

**Comment by amirouche** [comment]
@alex-eri it doesn't work well outside production, except if you also **have** to configure **nginx** on the development host.

**Comment by amirouche** [comment]
I am willing to work on this.

> I mean we could add a method `router.add_subdomain('sub.domain.com', subapp)` like recently added `router.add_subapp('/prefix', subapp)`.

And prolly:

> wildcards in the domain name as well a falback so unknown domains are handled too

**Comment by amirouche** [comment]
To be perfect, aiohttp needs a replacement for [eoe](https://pypi.org/project/eoe/):

> `*.hostname.local` dns resolver, [pow.cx](http://pow.cx/) replacement wanna-be.

**Comment by asvetlov** [comment]
> To be perfect, aiohttp needs a replacement for eoe:

Sorry, I don't follow

**Comment by amirouche** [comment]
Sorry, please forget about the eoe thing.

**Comment by mlyko** [comment]
Hi there,

I'd like to ask you if you going to add this feature to the aiohttp official release at all? I'm asking because the corresponding pull request seems to be abandoned by an author.

And second question. Why don't you want to add an ordinary routing by the 'Host' header to aiohttp, like it is available almost in all modern web framework (e.g. tornado, sanic)? Like this:
`
router.add_route('GET', '/path/', handler, host='example.com')
`

I could provide an appropriate pull request.

**Comment by asvetlov** [comment]
Currently nobody works on the feature AFAIK but we very appreciate the contribution.

Regarding to second question (`router.add_route('GET', '/path/', handler, host='example.com')`).
I think it is not necessary: the code mixes routing by host and bare requests.
Maybe better to support explicit sub-app with host (domain) specified.

**Comment by mlyko** [comment]
Sorry, but I don't get it, what do you mean writing that 'the code mixes routing by host and bare requests'?
Please take a look at a concept of such routing which I've already implemented:
https://github.com/mlyko/aiohttp/commit/9054d0b6f2ab69b42d80e5f6d8642a19e6d09382
and give a feedback.

I don't like the concept with sub-apps in this case because I'll have to create sub-apps to use the routing by host feature. Whereas sub-apps in aiohttp are for solving problem with big projects, not for routing purposes.

**Comment by mlyko** [comment]
Ok, thanks for clarifying. So it means that futher developing an implementation of routing by host proposed by me doesn't make any sense :)

**Comment by asvetlov** [comment]
But explicit domain/subdomain support does make sense :)

**Comment by roganov** [comment]
It's probably a little tricky, but route and subdomain rules can be combined in a single regular expression. I believe werkzeug does this.

**Comment by asvetlov** [comment]
I pretty sure the combining is an implementation artifact, not the main intention.
As an alternative we can consider NGINX (or Apache) config. The config has `server` and `location` directives, pretty easy and understandable.

**Comment by amirouche** [comment]
@asvetlov Can you summarize you point of view regarding this issue and my pull request https://github.com/amirouche/aiohttp/commit/4b0c9ff2d0895880a6f6f2966dd273b3584984ee. Should I continue to work on this?

**Comment by samuelcolvin** [comment]
I believe this should have been closed by #2809
